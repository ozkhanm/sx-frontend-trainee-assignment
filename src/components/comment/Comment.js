import React, {useState} from "react";

import {getDate, createMarkup} from "../../utils";

const Comment = (props) => {
    const {comment} = props;
    const [isButtonPressed, setIsButtonPressed] = useState(false);
    const [kidsCommentsParentId, setKidsCommentsParentId] = useState(-1);

    const getButtonElement = (comment) => {
        const id = comment.id;

        if ((!isButtonPressed) && (comment.hasOwnProperty(`kids`))) {
            return (
                <button className="comment-show-more-block" type="button" onClick={() => {
                    setKidsCommentsParentId(id);
                    setIsButtonPressed(true);
                }}>show more</button>
            );
        } else {
            return;
        }
    };

    const getCommentBlock = (comment) => {
        return (
            <div className="comment-parent-block">
                <div className="comment-item-header-block">
                    <p className="comment-item">by: {comment.by}</p>
                    <p className="comment-item">at: {getDate(comment.time)}</p>
                </div>
                <p className="comment-item-text" dangerouslySetInnerHTML={createMarkup(comment.text)}></p>
            </div>
        );
    };

    const getSubcomments = (comments) => {
        return comments.map((it) => {
            if ((!it.hasOwnProperty(`deleted`)) && (!it.hasOwnProperty(`dead`))) {
                return (
                    <div key={it.id} className="comment-item-block">
                        {getCommentBlock(it)}
                        {it.hasOwnProperty(`kids`) ? getSubcomments(it.kids) : null}
                    </div>
                ); 
            } else {
                return;
            }
        });
    };

    const getSubcommentElement = (id) => {
        if (kidsCommentsParentId === id) {
            const comments = comment.kids;
            
            return (
                <React.Fragment>
                    {getSubcomments(comments)}
                </React.Fragment>
            );
        }
    };

    const getParentCommentElement = (comment) => {
        return (
            <div className="comment-item-block">
                {getCommentBlock(comment)}
                {getButtonElement(comment)}
                {getSubcommentElement(comment.id)}
            </div>
        );
    };

    if ((!comment.hasOwnProperty(`deleted`)) && (!comment.hasOwnProperty(`dead`))) {
        return getParentCommentElement(comment);
    } else {
        return null;
    }
};

export default Comment;