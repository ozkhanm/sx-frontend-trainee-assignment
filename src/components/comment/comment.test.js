import React from "react";
import renderer from "react-test-renderer";

import Comment from "./Comment";

it(`Comment component without kids renders correctly`, () => {
    const comment = {
        by: "vbphprubyjsgo",
        id: 30387743,
        parent: 30386645,
        text: "Isn&#x27;t that just a techspam website? I only read their monitor reviews and they were all wrong.",
        time: 1645201625,
        type: "comment"
    };
    const tree = renderer
        .create(<Comment comment={comment}/>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});

it(`Comment component with kids renders correctly`, () => {
    const comment = {
        by: "awill",
        id: 30389916,
        kids: [{
            by: "IanCutress",
            id: 30390027,
            parent: 30389916,
            text: "Simply understaffed. Since I&#x27;ve been full time, we&#x27;ve only ever had two full time employees.<p>Also, Samsung doesn&#x27;t sample AnandTech.",
            time: 1645212533,
            type: "comment"
        }],
        parent: 30386645,
        text: "I feel AT is great for enthusiasts. I want them to become more mainstream, as their reviews are so much better than generic tech sites, but they need a better balance. They sometimes release iPhone, Galaxy or Pixel reviews weeks or even months after launch. Just look at Youtube, and impact on views from being first with a review. Reviews that arrive weeks after a product launch don&#x27;t have mass-market appeal. AT either doesn&#x27;t understand the market, or is understaffed.<p>I get a thorough review can&#x27;t come out on day one, but they need to get something out. Maybe a first look with some opinions, followed by a more thorough review a week later.",
        time: 1645211797,
        type: "comment"
    };
    const tree = renderer
        .create(<Comment comment={comment}/>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});