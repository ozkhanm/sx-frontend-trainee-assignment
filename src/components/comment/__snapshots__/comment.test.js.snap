// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component with kids renders correctly 1`] = `
<div
  className="comment-item-block"
>
  <div
    className="comment-parent-block"
  >
    <div
      className="comment-item-header-block"
    >
      <p
        className="comment-item"
      >
        by: 
        awill
      </p>
      <p
        className="comment-item"
      >
        at: 
        18.02.2022, 22:16
      </p>
    </div>
    <p
      className="comment-item-text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "I feel AT is great for enthusiasts. I want them to become more mainstream, as their reviews are so much better than generic tech sites, but they need a better balance. They sometimes release iPhone, Galaxy or Pixel reviews weeks or even months after launch. Just look at Youtube, and impact on views from being first with a review. Reviews that arrive weeks after a product launch don&#x27;t have mass-market appeal. AT either doesn&#x27;t understand the market, or is understaffed.<p>I get a thorough review can&#x27;t come out on day one, but they need to get something out. Maybe a first look with some opinions, followed by a more thorough review a week later.",
        }
      }
    />
  </div>
  <button
    className="comment-show-more-block"
    onClick={[Function]}
    type="button"
  >
    show more
  </button>
</div>
`;

exports[`Comment component without kids renders correctly 1`] = `
<div
  className="comment-item-block"
>
  <div
    className="comment-parent-block"
  >
    <div
      className="comment-item-header-block"
    >
      <p
        className="comment-item"
      >
        by: 
        vbphprubyjsgo
      </p>
      <p
        className="comment-item"
      >
        at: 
        18.02.2022, 19:27
      </p>
    </div>
    <p
      className="comment-item-text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Isn&#x27;t that just a techspam website? I only read their monitor reviews and they were all wrong.",
        }
      }
    />
  </div>
</div>
`;
